fuzz: improve `coinselection`

brunoerg opened the pull request:
This PR:

- Moves coin creation to its own function called `CreateCoins`.
- Add coverage for `EligibleForSpending`
- Add coverage for `AddInputs`: get result of each algorithm (srd, knapsack and bnb), call `CreateCoins` and add into them.
- Add coverage for `GetShuffledInputVector` and `GetInputSet` using the result of each algorithm (srd, knapsack and bnb).
- Add coverage for `Merge`: Call SRD with the new utxos and, if successful, try to merge with the previous SRD result.

brunoerg committed:
fuzz: coinselection, add `CreateCoins`

Move coins creation for a specific function. It
allows us to use it in other parts of the code.

brunoerg committed:
fuzz: coinselection, add coverage for `EligibleForSpending`

brunoerg committed:
fuzz: coinselection, add coverage for `AddInputs`

brunoerg committed:
fuzz: coinselection, add coverage for `GetShuffledInputVector`/`GetInputSet`

brunoerg committed:
fuzz: coinselection, add coverage for `Merge`

fanquake commented:
cc @achow101 @Xekyo @furszy

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -114,6 +114,19 @@ FUZZ_TARGET(coinselection)
     }
+
+    std::vector<COutput> utxos;
+    std::set<std::shared_ptr<COutput>> utxos_set;
+    CreateCoins(fuzzed_data_provider, utxos, coin_params);
+    std::transform(utxos.begin(), utxos.end(),
+                 std::inserter(utxos_set, utxos_set.begin()),
+                 [](const COutput& output) {
+                   return std::make_shared<COutput>(output);
+                 });
+
+    if (result_srd) result_srd->AddInputs(utxos_set, subtract_fee_outputs);
+    if (result_knapsack) result_knapsack->AddInputs(utxos_set, subtract_fee_outputs);
+    if (result_bnb) result_bnb->AddInputs(utxos_set, subtract_fee_outputs);
```
If `CreateCoins` generates a coin that is already inside any of the results, `AddInputs` will throw a runtime error, which the fuzzing framework will detect as a failure when it is not. It is part of the class boundaries.

Same apply for the fuzz `Merge` function commit.

furszy replied:
No ok. The reason why this doesn't fails is the test global increasing `next_locktime` which ensures that all UTXOs receive a different hash even when they have the same data.
So, nvm. A comment wouldn't hurt but nothing serious.

brunoerg replied:
Got it, thanks.

murchandamus reviewed (commented):
I’m pretty new to fuzz testing, so I am not sure whether we want to have all these additional calls for each selection result. It seems to me that if we wanted to test the behavior of SelectionResult, it might make more sense to separately fuzz that instead of fuzzing that in the context of running all the different coin selection algorithms, since it might lower the number of executions we get on fuzzing coin selection.

OTOH, it’s great to get fuzzing coverage for these codepaths.

Code changes look good to me. 

Tentative ACK d5282a45bd452844dc815181825e3dd46133ac51, except that I’d be curious to hear what others think about expanding the responsibility of this fuzz target vs adding a second fuzz target for testing SelectionResult.

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -68,40 +87,65 @@ FUZZ_TARGET(coinselection)
     auto result_srd = SelectCoinsSRD(group_pos, target, fast_random_context, MAX_STANDARD_TX_WEIGHT);
-    if (result_srd) result_srd->ComputeAndSetWaste(cost_of_change, cost_of_change, 0);
+    if (result_srd) {
+        result_srd->ComputeAndSetWaste(cost_of_change, cost_of_change, 0);
```
I didn't touch `ComputeAndSetWaste` parameters value, perhaps who did it copied (?) from: https://github.com/bitcoin/bitcoin/blob/34ac3f438a268e83af6cd11e2981e5bc07f699c9/src/wallet/coinselection.cpp#L179

I'm going to change it to make it more "realistic"/robust.

brunoerg replied:
Done

brunoerg committed:
fuzz: coinselection, improve `ComputeAndSetWaste`

Instead of using `cost_of_change` for `min_viable_change`
and `change_cost`, and 0 for `change_fee`, use values from
`coin_params`. The previous values don't generate any effects
that is relevant for that context.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1213485109 and https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1213479178

Xekyo commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -64,44 +84,71 @@ FUZZ_TARGET(coinselection)
     auto result_srd = SelectCoinsSRD(group_pos, target, fast_random_context, MAX_STANDARD_TX_WEIGHT);
-    if (result_srd) result_srd->ComputeAndSetWaste(cost_of_change, cost_of_change, 0);
+    if (result_srd) {
+        result_srd->ComputeAndSetWaste(coin_params.min_viable_change, coin_params.m_cost_of_change, coin_params.m_change_fee);
```
I think you might get some funny behavior here, if `min_viable_change` is smaller than `m_change_fee` as mentioned above, also see https://github.com/bitcoin/bitcoin/pull/27846

Xekyo replied:
Nevermind that. I ran over about 500k executions on the coinselection fuzz test and could not substantiate my concern.

murchandamus reviewed (approved):
ACK 2d739a425c92e2dd409c45273e8e376d39ead1cc

My concern about the `min_viable_change` and `change_fee` causing issues did not substantiate after some light fuzzing. Since it does not seem to cause crashes, having independent values for the two parameters may be better. Thanks for adding more coverage.

brunoerg commented:
Rebased

murchandamus reviewed (commented):
reACK bd6315a2a75d2891a8d3476c492d6b53309b7296

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -117,6 +117,19 @@ FUZZ_TARGET(coinselection)
         assert(result_knapsack);
     }
+
+    std::vector<COutput> utxos;
+    std::set<std::shared_ptr<COutput>> utxos_set;
+    CreateCoins(fuzzed_data_provider, utxos, coin_params);
+    std::transform(utxos.begin(), utxos.end(),
+                 std::inserter(utxos_set, utxos_set.begin()),
+                 [](const COutput& output) {
+                   return std::make_shared<COutput>(output);
+                 });
```
nit: To not have to do the transformation, could move all the `std::vector<COutput>` to `std::vector<std::shared_ptr<COutput>>`

brunoerg replied:
Done

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -101,16 +101,23 @@ FUZZ_TARGET(coinselection)
     auto result_bnb = SelectCoinsBnB(group_pos, target, cost_of_change, MAX_STANDARD_TX_WEIGHT);
+    if (result_bnb) {
+        (void)result_bnb->GetShuffledInputVector();
+    }
```
Couldn't we assert that BnB `GetChange()` is always 0?

brunoerg replied:
Does it worth? BnB internally calls `ComputeSetAndWaste` which checks change. Any unexpected action regarding it would affect the assertion - `assert(best_waste == result.GetWaste());`.

furszy replied:
I think that it worth. One of the fuzzing goals is to assert expected behavior externally. Black-box style. So if the internal implementation changes by mistake (or if there is a way to by-pass the expected behavior that we aren't aware of), the fuzzer catches it.

Also, the internal BnB `ComputeSetAndWaste` call checks that the final selection is within a certain range: `target <= value <  target + change cost`. It is indirectly checking `GetChange()` result but I don't think that we should rely on that. If `change_cost` is also 0, the `GetChange` result doesn't really matter, the same code will be executed in both branches of the `ComputeAndSetWaste` function.

brunoerg replied:
Fair enough, gonna address it.

brunoerg replied:
Done!

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -140,6 +140,13 @@ FUZZ_TARGET(coinselection)
+    std::vector<OutputGroup> new_group_pos;
+    GroupCoins(fuzzed_data_provider, utxos, coin_params, /*positive_only=*/true, new_group_pos);
+    auto new_srd = SelectCoinsSRD(new_group_pos, target, coin_params.m_change_fee, fast_random_context, MAX_STANDARD_TX_WEIGHT);
+    if (new_srd && result_srd) {
+        new_srd->Merge(result_srd.value());
+    }
```
In d72c0926:

Should check the `Merge` outcome. e.g. the result target and weight need to be the sum of the two merged targets and weights. Moreover, if one result uses the effective value and the other one not, the `use_effective` field must be updated.

brunoerg replied:
Done

furszy reviewed (commented):
Left few comments. I think that we could go further in few places and check that the called functions are actually doing what they suppose to be doing.

For instance, all valid results must have a target below the sum of the selected inputs amounts. Also, waste on results with more difference between target and inputs sum should be greater than results with closer diff between target and inputs sum.

brunoerg committed:
fuzz: coinselection, compare `GetSelectedValue` with target

The valid results should have a target below the sum of
the selected inputs amounts. Also, it increases the
minimum value for target to make it more realistic.

brunoerg commented:
Force-pushed:

Addressed:
- https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1272359281
- https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1272371310

Also, I added support for `GetSelectedValue()`, asserting that its value is always greater or equal than `target` in a valid result.

brunoerg committed:
fuzz: coinselection, BnB should never produce change

brunoerg commented:
Force-pushed addressing @furszy's suggestion to check whether `GetChange` is always 0 for any BnB result.

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -118,6 +118,17 @@ FUZZ_TARGET(coinselection)
+        utxos_set.insert(std::make_shared<COutput>(utxo));
+    }
+
+    if (result_srd) result_srd->AddInputs(utxos_set, subtract_fee_outputs);
+    if (result_knapsack) result_knapsack->AddInputs(utxos_set, subtract_fee_outputs);
+    if (result_bnb) result_bnb->AddInputs(utxos_set, subtract_fee_outputs);
```
what about asserting inputs are actually added and checking the total weight?

brunoerg replied:
Sounds good. Just did it.

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -77,7 +77,7 @@ FUZZ_TARGET(coinselection)
     const CAmount cost_of_change{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
     const CAmount min_viable_change{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
-    const CAmount target{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY)};
+    const CAmount target{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(50000, MAX_MONEY)};
```
what was the reason behind this change?

brunoerg replied:
just changed to `1000`. but the reason is to make it more realistic. e.g. do we really use it for 1 sat?

furszy reviewed (commented):
Code review ACK 1e0f5f088.

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -139,6 +139,17 @@ FUZZ_TARGET(coinselection)
+        new_srd->Merge(result_srd.value());
+        assert(new_srd->GetTarget() == old_target + result_srd->GetTarget());
+        assert(new_srd->GetWeight() == old_weight + result_srd->GetWeight());
+    }
```
Just changed it to not use two srd solutions and I created a function to "simulate" a manual selection. Also, added more asserts, including checking the input set.

brunoerg commented:
Force-pushed:

- Improved `Merge`: now it checks the input set and it tries to merge every solution with a "manual" one.
- Improved `AddInputs` to check the total weight.

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -144,6 +155,21 @@ FUZZ_TARGET(coinselection)
+    CAmount manual_balance{CreateCoins(fuzzed_data_provider, manual_utxos, coin_params)};
+    if (manual_balance <= 0) return;
+    for (auto& result : results) {
+        if (!result) break;
```
`continue`?

brunoerg replied:
yea, done!

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -118,6 +118,22 @@ FUZZ_TARGET(coinselection)
+            utxos_set.insert(std::make_shared<COutput>(utxo));
+        }
+        for (auto& result : results) {
+            if (!result) break;
```
`continue`?

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -118,6 +118,22 @@ FUZZ_TARGET(coinselection)
+            if (!result) break;
+            const auto weight{result->GetWeight()};
+            result->AddInputs(utxos_set, subtract_fee_outputs);
+            assert(result->GetWeight() > weight);
```
nit: could verify the exact weight by summing up all the utxos weight.

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -144,6 +155,21 @@ FUZZ_TARGET(coinselection)
+
+    std::vector<COutput> manual_utxos;
+    CAmount manual_balance{CreateCoins(fuzzed_data_provider, manual_utxos, coin_params)};
+    if (manual_balance <= 0) return;
```
can't be negative. It's a ranged amount.

brunoerg replied:
Done

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -144,6 +155,21 @@ FUZZ_TARGET(coinselection)
+    if (manual_balance <= 0) return;
+    for (auto& result : results) {
+        if (!result) break;
+        auto manual_selection{ManualSelection(manual_utxos, manual_balance, coin_params.m_subtract_fee_outputs)};
```
The manual selection is always the same. Could place it outside of the loop.

brunoerg replied:
Done

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -144,6 +155,21 @@ FUZZ_TARGET(coinselection)
+        const std::set<std::shared_ptr<COutput>> input_set{result->GetInputSet()};
+        const int old_weight{result->GetWeight()};
+        result->Merge(manual_selection);
+        assert(result->GetInputSet().size() > input_set.size());
```
```c++
assert(result->GetInputSet().size() == input_set.size() + manual_utxos.size());
```

brunoerg replied:
Done

brunoerg commented:
Force-pushed addressing:
- https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1289387991
- https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1289387061
- https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1289386461
- https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1289382284

Thanks, @furszy!

furszy reviewed (commented):
utACK 1942ea0

murchandamus commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -88,7 +88,7 @@ FUZZ_TARGET(coinselection)
     const CFeeRate effective_fee_rate{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
     const CAmount cost_of_change{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
     const CAmount min_viable_change{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
-    const CAmount target{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY)};
```
While we don’t broadcast transactions with outputs of that amount (except OP_RETURN maybe), if it doesn’t break anywhere, a target of 1 ṩ should maybe be legal for coin selection?

IIRC, Bitcoin Core permits 294 ṩ for P2WPKH outputs, so it should at least not be bigger than that.

brunoerg replied:
I checked that 1ṩ  may break `assert(result_bnb->GetChange(coin_params.m_cost_of_change, CAmount{0}) == 0);`

murchandamus replied:
Ah, my bad, I thought this target referred to the sum of recipient outputs, but it’s actually `recipient_sum + not_input_fees`.

Is it possible that this then would need to be at least as large as the cost for the header + change output? 1000 ṩ/vB might not be enough for every feerate then.

brunoerg replied:
Nice, will leave it running tonight with some values (1000, 5000, etc) to check it as well.

murchandamus replied:
I got a crash after 7.4M attempts:

    fuzz: ../../src/wallet/test/fuzz/coinselection.cpp:120: void wallet::coinselection_fuzz_target(FuzzBufferType): Assertion `result_bnb->GetChange(coin_params.m_cost_of_change, CAmount{0}) == 0' failed.`
    
You should be able to replicate the issue with:
    
$ echo "MP3//////wT/LzMBEABL////////Wv///yWyEABLAADoQP//PP//CAAAPQAAAAAAAAAAPQEAAAAA AAAA/V12+w==" | base64 -d > crash.input 
$ FUZZ=coinselection src/test/fuzz/fuzz crash.input

murchandamus reviewed (approved):
Please lower the minimum target, otherwise 1942ea0e44c8b4143c9d97f23ff31dc9a6f64a56 looks ready to me. Nits optional of course. :)

murchandamus commented:
The fuzz test crashed for me after 7.4M attempts:

    fuzz: ../../src/wallet/test/fuzz/coinselection.cpp:120: void wallet::coinselection_fuzz_target(FuzzBufferType): Assertion `result_bnb->GetChange(coin_params.m_cost_of_change, CAmount{0}) == 0' failed.`
    
You should be able to replicate the issue with:
    
```
$ echo "MP3//////wT/LzMBEABL////////Wv///yWyEABLAADoQP//PP//CAAAPQAAAAAAAAAAPQEAAAAA AAAA/V12+w==" | base64 -d > crash.input 
$ FUZZ=coinselection src/test/fuzz/fuzz crash.input
```

brunoerg commented:
@murchandamus I got a crash with 5000 as well, perhaps we should increase the value even more.

murchandamus commented:
I don’t think that an absolute value will work. My guess is that the crash may be caused by `m_cost_of_change` being generated randomly independent from `m_change_fee`. Usually, `cost_of_change` cannot ever be smaller than `m_change_fee`, so if it happens to be smaller here, it may interfere with the change calculation?

 I attempted to calculate  `cost_of_change` instead from `m_change_fee + input_size×LTFRE`, leaving the fuzz input consumption unchanged. It made the test pass:

```diff
@@ -97,9 +97,9 @@ FUZZ_TARGET(coinselection)
     coin_params.m_long_term_feerate = long_term_fee_rate;
     coin_params.m_effective_feerate = effective_fee_rate;
     coin_params.min_viable_change = min_viable_change;
-    coin_params.m_cost_of_change = cost_of_change;
     coin_params.change_output_size = fuzzed_data_provider.ConsumeIntegralInRange<int>(10, 1000);
     coin_params.m_change_fee = effective_fee_rate.GetFee(coin_params.change_output_size);
+    coin_params.m_cost_of_change = coin_params.m_change_fee +  long_term_fee_rate.GetFee(58);
```

brunoerg commented:
> I don’t think that an absolute value will work. My guess is that the crash may be caused by m_cost_of_change being generated randomly independent from m_change_fee. Usually, cost_of_change cannot ever be smaller than m_change_fee, so if it happens to be smaller here, it may interfere with the change calculation?

Bingo, I think so!

murchandamus commented:
That was loosely inspired by the [vsize of a P2TR input](https://bitcoin.stackexchange.com/q/111395/5406) (rounded up). 

I’m just seeing that we are calculating the cost_of_change actually with the `discard_feerate` rather than the `LTFRE`, though:

`src/wallet/spend.cpp:   coin_selection_params.m_cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + coin_selection_params.m_change_fee;`

You could then also make the `change_spend_size` or `change_input_size` a random variable in the range `41…1000`. Inputs cannot ever be smaller than 41 vB (32+4+4+1).

brunoerg committed:
fuzz: coinselection, fix `m_cost_of_change`

`m_cost_of_change` must not be generated randomly
independent from m_change_fee. This commit changes
it to set it up according to `wallet/spend`.

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -65,6 +65,17 @@ static CAmount CreateCoins(FuzzedDataProvider& fuzzed_data_provider, std::vector
+static SelectionResult ManualSelection(std::vector<COutput>& utxos, const CAmount& total_amount, const bool& subtract_fee_outputs)
+{
+    SelectionResult result(total_amount, SelectionAlgorithm::MANUAL);
+    std::set<std::shared_ptr<COutput>> utxos_set;
+    for (const auto& utxo : utxos) {
+        utxos_set.insert(std::make_shared<COutput>(utxo));
+    }
+    result.AddInputs(utxos_set, subtract_fee_outputs);
```
Done!

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -144,6 +155,21 @@ FUZZ_TARGET(coinselection)
         }
     }
+
+    std::vector<COutput> manual_utxos;
```
Done!

brunoerg commented:
Force pushed addressing https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1293959610, https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1293962271, https://github.com/bitcoin/bitcoin/pull/27585#discussion_r1293962878 and changed the `cost_of_change` according to `spend.cpp`, it avoids any issue with low targets.

murchandamus reviewed (commented):
Code changes look good to me, crACK d501e1a2283d82b833260bf3372512d616a62a4c

Started some fuzzing, will let you know if it finds anything.

murchandamus commented:
My fuzzer has performed over 10M runs without uncovering an issue.

ACK https://github.com/bitcoin/bitcoin/commit/d501e1a2283d82b833260bf3372512d616a62a4c

furszy commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -45,6 +45,37 @@ static void GroupCoins(FuzzedDataProvider& fuzzed_data_provider, const std::vect
     if (valid_outputgroup) output_groups.push_back(output_group);
 }
 
+static CAmount CreateCoins(FuzzedDataProvider& fuzzed_data_provider, std::vector<COutput>& utxo_pool, CoinSelectionParams& coin_params)
+{
+    // Create some coins
+    CAmount total_balance{0};
+    int next_locktime{0};
```
Need to make `next_locktime` a global variable (or receive it as ref here).

This variable prevents `CreateCoins` from generating duplicated transaction hashes (look at `AddCoin()`). If the variable is reset on every `CreateCoins`, as is now, could be the case where the same coins are created twice.
Which will cause a runtime_error in `AddInputs()` and `Merge()` calls (see `SelectionResult::InsertInputs()`).

brunoerg replied:
Done! Thanks!

brunoerg commented:
Force-pushed fixing `next_locktime` in `CreateCoins`. Thanks, @furszy

furszy reviewed (commented):
Code review ACK 5f0237f2

As an extra small nit, I would merge the `GetInputSet` and `GetShuffledInputSet` commits into one (ec82d1c3a523a9b28adf66f840e239f4adead37a, 4531ab1aee85a6801e0de00816b3764b8462771c) as they are basically doing the same. But, not a big deal if you don't do it anyway.

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -84,7 +84,8 @@ FUZZ_TARGET(coinselection)
     const CFeeRate effective_fee_rate{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
-    const CAmount cost_of_change{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
+    // Discard feerate must be at least dust relay feerate
+    const CFeeRate discard_fee_rate{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(3000, COIN)};
```
Done

brunoerg commented:
Force-pushed: use `DUST_RELAY_TX_FEE` instead of hardcoding the value and squashed `GetInputSet` and `GetShuffledInputSet` commits.

furszy reviewed (commented):
re-ACK bf26f97

murchandamus commented:
reACK with some minimal fuzzing bf26f978ffbe7e2fc681825de631600e24e5c93e

achow101 commented:
ACK bf26f978ffbe7e2fc681825de631600e24e5c93e

fanquake commented:
> This crashes:

Yea. OSS-Fuzz has this this now as well https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=61790:
```bash
Crash Type: ASSERT
Crash Address: 
Crash State:
  result_bnb->GetChange(coin_params.m_cost_of_change, CAmount{0}) == 0
  wallet::coinselection_fuzz_target
  std::__1::__function::__func<void
```

brunoerg commented:
Investigating.

murchandamus commented:
Follow-up in #28372

brunoerg commented:
Fuzz found an issue! Fix #28395
